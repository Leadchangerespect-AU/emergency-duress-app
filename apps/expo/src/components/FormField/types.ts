export interface FormFieldProps {
  label: string;
  isRequired?: boolean;
  isDisabled?: boolean;
  errors: (undefined | false | null | string)[];
}
