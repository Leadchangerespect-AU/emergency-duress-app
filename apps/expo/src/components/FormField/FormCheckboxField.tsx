import {
  FormControl,
  FormControlError,
  FormControlErrorIcon,
  FormControlErrorText,
  FormControlLabel,
} from "@/components/FormControl";
import Checkbox, { CheckboxProps } from "../Checkbox";
import { FormFieldProps } from "./types";

export const FormCheckboxField = ({
  isRequired,
  errors,
  isDisabled,
  children,
  ...props
}: Omit<FormFieldProps, "label"> & CheckboxProps) => {
  const hasErrors = errors.length > 0;

  return (
    <FormControl
      isInvalid={hasErrors}
      isRequired={isRequired}
      isDisabled={isDisabled}
    >
      <Checkbox {...props}>{children}</Checkbox>
      {hasErrors && (
        <FormControlError>
          <FormControlErrorIcon />
          <FormControlErrorText>{errors.join(", ")}</FormControlErrorText>
        </FormControlError>
      )}
    </FormControl>
  );
};
