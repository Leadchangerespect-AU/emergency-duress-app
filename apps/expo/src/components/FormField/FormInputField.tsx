import { useId } from "react";
import { Platform, TextInputProps } from "react-native";

import {
  FormControl,
  FormControlError,
  FormControlErrorIcon,
  FormControlErrorText,
  FormControlLabel,
} from "@/components/FormControl";
import Input from "@/components/Input";
import { FormFieldProps } from "./types";

export const FormInputField = ({
  label,
  isRequired,
  isDisabled,
  errors,
  ...props
}: FormFieldProps & TextInputProps) => {
  const hasErrors = errors.length > 0;

  const labelId = useId();
  const inputId = useId();
  return (
    <FormControl
      isInvalid={hasErrors}
      isRequired={isRequired}
      isDisabled={isDisabled}
    >
      <FormControlLabel
        id={labelId}
        {...Platform.select({
          web: {
            htmlFor: inputId,
          },
        })}
      >
        {label}
      </FormControlLabel>
      <Input
        {...props}
        aria-labelledby={labelId}
        accessibilityLabel={label}
        id={inputId}
      />
      {hasErrors && (
        <FormControlError>
          <FormControlErrorIcon />
          <FormControlErrorText>{errors.join(", ")}</FormControlErrorText>
        </FormControlError>
      )}
    </FormControl>
  );
};
