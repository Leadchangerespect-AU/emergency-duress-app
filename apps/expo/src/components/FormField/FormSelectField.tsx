import { useId } from "react";
import { Platform } from "react-native";

import {
  FormControl,
  FormControlError,
  FormControlErrorIcon,
  FormControlErrorText,
  FormControlLabel,
} from "@/components/FormControl";
import Select from "@/components/Select";
import { SelectProps } from "@/components/Select/types";
import { FormFieldProps } from "./types";

export const FormSelectField = ({
  label,
  isRequired,
  isDisabled,
  errors,
  ...props
}: FormFieldProps & SelectProps) => {
  const hasErrors = errors.length > 0;

  const labelId = useId();
  const selectId = useId();
  return (
    <FormControl
      isInvalid={hasErrors}
      isRequired={isRequired}
      isDisabled={isDisabled}
    >
      <FormControlLabel
        id={labelId}
        {...Platform.select({
          web: {
            htmlFor: selectId,
          },
        })}
      >
        {label}
      </FormControlLabel>
      <Select
        {...props}
        aria-labelledby={labelId}
        accessibilityLabel={label}
        id={selectId}
      />
      {hasErrors && (
        <FormControlError>
          <FormControlErrorIcon />
          <FormControlErrorText>{errors.join(", ")}</FormControlErrorText>
        </FormControlError>
      )}
    </FormControl>
  );
};
