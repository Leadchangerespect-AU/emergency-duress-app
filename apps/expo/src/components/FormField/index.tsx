import { FormInputField } from "./FormInputField";
import { FormSelectField } from "./FormSelectField";

export { FormInputField, FormSelectField };
