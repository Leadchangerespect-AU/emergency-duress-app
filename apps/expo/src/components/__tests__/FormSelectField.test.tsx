import React from "react";
import { Platform } from "react-native";
import { fireEvent, render, screen } from "@testing-library/react-native";

import { FormSelectField } from "../FormField/FormSelectField";

// Mock the Ionicons import used by FormControlErrorIcon
jest.mock("@expo/vector-icons", () => ({
  Ionicons: "",
}));

describe("FormSelectField", () => {
  const defaultProps = {
    label: "Test Select",
    options: [
      { label: "Option 1", value: "1" },
      { label: "Option 2", value: "2" },
      { label: "Option 3", value: "3" },
    ],
    value: null,
    onChange: jest.fn(),
    errors: [],
  };

  it("renders with required label", () => {
    render(<FormSelectField {...defaultProps} isRequired />);
    const input = screen.getByLabelText("Test Select");
    expect(input).toBeTruthy();
    expect(input.props.accessibilityLabel).toBe("Test Select");
  });

  it("displays error messages", () => {
    const errors = ["Required field", "Invalid selection"];
    render(<FormSelectField {...defaultProps} errors={errors} />);

    expect(screen.getByText("Required field, Invalid selection")).toBeTruthy();
  });

  it("handles selection changes", () => {
    const onChange = jest.fn();
    render(
      <FormSelectField
        {...defaultProps}
        onChange={onChange}
        testID="form-select"
      />,
    );

    const select = screen.getByTestId("form-select");
    fireEvent(select, "onChange", { target: { value: "2" } });
    expect(onChange).toHaveBeenCalledWith({ target: { value: "2" } });
  });

  it("handles web platform behavior", () => {
    Platform.OS = "web";
    render(<FormSelectField {...defaultProps} testID="form-select" />);

    const select = screen.getByTestId("form-select");
    expect(select).toBeTruthy();
  });

  it("handles native platform behavior", () => {
    Platform.OS = "ios";
    render(<FormSelectField {...defaultProps} testID="form-select" />);

    const select = screen.getByTestId("form-select");
    expect(select).toBeTruthy();
  });

  it("matches snapshot", () => {
    render(<FormSelectField {...defaultProps} />);
    expect(screen.toJSON()).toMatchSnapshot();
  });
});
