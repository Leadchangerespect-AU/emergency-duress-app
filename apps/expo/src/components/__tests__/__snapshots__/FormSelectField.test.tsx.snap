// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormSelectField matches snapshot 1`] = `
<View
  className="my-2"
>
  <Text
    className="text-base text-neutral-800 font-semibold"
    id=":ra:"
  >
    Test Select
  </Text>
  <View
    accessibilityLabel="Test Select"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    aria-labelledby=":ra:"
    className="rounded border bg-white px-4 py-2"
    collapsable={false}
    focusable={true}
    id=":rb:"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <Text
      className="text-black"
    >
      Select...
    </Text>
  </View>
</View>
`;
